src/plugins/part.c

```diff
@@ -2100,85 +2065,150 @@ gboolean bd_part_set_disk_flag (const gchar *disk, BDPartDiskFlag flag, gboolean
-            if ((1 << i) & flags)
-                status = ped_partition_set_flag (ped_part, (PedPartitionFlag) i, (int) 1);
-            else if (ped_partition_is_flag_available (ped_part, (PedPartitionFlag) i))
-                status = ped_partition_set_flag (ped_part, (PedPartitionFlag) i, (int) 0);
-            if (status == 0) {
-                set_parted_error (error, BD_PART_ERROR_FAIL);
-                g_prefix_error (error, "Failed to set flag on the partition '%d' on device '%s'", part_num, disk);
-                ped_disk_destroy (ped_disk);
-                ped_device_destroy (dev);
-                bd_utils_report_finished (progress_id, (*error)->message);
-                return FALSE;
+        if (!set_part_type (cxt, part_num, DEFAULT_PART_ID, BD_PART_TABLE_MSDOS, error)) {
+            g_prefix_error (error, "Failed to reset partition ID on partition '%s': ", part);
+            bd_utils_report_finished (progress_id, (*error)->message);
+            close_context (cxt);
+            return FALSE;
+        }
+    } else if (table_type == BD_PART_TABLE_GPT) {
+        if (!set_part_type (cxt, part_num, DEFAULT_PART_GUID, BD_PART_TABLE_GPT, error)) {
+            g_prefix_error (error, "Failed to reset partition type on partition '%s': ", part);
+            bd_utils_report_finished (progress_id, (*error)->message);
+            close_context (cxt);
+            return FALSE;
+        }
+    }
+
+    /* and now go through all the flags and set the required ones */
+    if (table_type == BD_PART_TABLE_MSDOS) {
+        /* special cases first */
+        if (flags & BD_PART_FLAG_BOOT && !set_boot_flag (cxt, part_num, TRUE, error)) {
+            close_context (cxt);
+            g_prefix_error (error, "Failed to set boot flag on partition '%s': ", part);
+            bd_utils_report_finished (progress_id, (*error)->message);
+            return FALSE;
+        }
+
+        if (flags & BD_PART_FLAG_HIDDEN || flags & BD_PART_FLAG_LBA) {
+            part_id = get_lba_hidden_id (part, flags & BD_PART_FLAG_HIDDEN, flags & BD_PART_FLAG_LBA, TRUE, error);
+            if (part_id) {
+                if (!set_part_type (cxt, part_num, part_id, BD_PART_TABLE_MSDOS, error)) {
+                    bd_utils_report_finished (progress_id, (*error)->message);
+                    g_free (part_id);
+                    close_context (cxt);
+                    return FALSE;
+                }
+            } else {
+                if (*error) {
+                    g_prefix_error (error, "Failed to calculate partition ID to set: ");
+                    bd_utils_report_finished (progress_id, (*error)->message);
+                    close_context (cxt);
+                    return FALSE;
+                }
             }
         }
 
-        ret = disk_commit (ped_disk, disk, error);
+        /* flags that are actually partition IDs */
+        for (gint i = 1; i < log2i (BD_PART_FLAG_BASIC_LAST); i++) {
```

Comment:
No need to compute the `log2i` over and over.

Reply:
Fixed.